scripts/jenkins/Jenkinsfile

```diff
@@ -0,0 +1,515 @@
+    def args_str = ""
+    args_map.each { key, value ->
+        args_str += "-D$key=\"$value\" "
+    }
+    sh("cmake $args_str $directory")
+}
+
+def publishCoverage(String test_name) {
+    echo "Start publication of coverage data"
+    sh "mkdir -p ssh/coverage/ || true"
+    sh "mv build/coverage ssh/coverage/${test_name} || true"
+    sshPublisher(publishers: [
+      sshPublisherDesc(
+        configName: 'doc.libelektra.org',
+        transfers: [
+          sshTransfer(
+            sourceFiles: 'ssh/**',
+            removePrefix: 'ssh'
+          )
+      ])
+    ])
+    echo "Finish publication of coverage data"
+}
+
+def getUid() {
+    return sh(returnStdout: true, script: 'id -u').trim()
+}
+
+def getGid() {
+    return sh(returnStdout: true, script: 'id -g').trim()
+}
+
+def trackCoverage(do_track, cl) {
+    if(do_track) {
+        sh 'make coverage-start'
+    }
+    cl()
+    if(do_track) {
+        sh 'make coverage-stop'
+        sh 'make coverage-genhtml'
+    }
+}
+
+def withDockerEnv(test_name, image, cl) {
+    return [(test_name): {
+        stage(test_name) {
+            node(docker_node_label) {
+                docker.withRegistry('https://hub.libelektra.org',
+                                    'docker-hub-elektra-jenkins') {
+                    timeout(activity: true, time: 5, unit: 'MINUTES') {
+                        def cpu_count = cpuCount()
+                        withEnv(["MAKEFLAGS='-j${cpu_count+2} -l${cpu_count*2}'",
+                                 "CTEST_PARALLEL_LEVEL='${cpu_count+2}'"]) {
+                            echo "Staring ${env.STAGE_NAME} on ${env.NODE_NAME}"
+                            checkout scm
+                            docker.image(image_id(image)).inside { cl() }
+                        }
+                    }
+                }
+            }
+        }
+    }]
+}
+
+def cpuCount() {
+    return sh(returnStdout: true,
+              script: 'grep -c ^processor /proc/cpuinfo').trim() as Integer
+}
+
+def ctest(target = "Test") {
+    sh """ctest -j ${env.CTEST_PARALLEL_LEVEL} --force-new-ctest-process \
+            --output-on-failure --no-compress-output -T ${target}"""
+}
+
+def cmemcheck(kdbtests) {
+    if(kdbtests) {
+        ctest("MemCheck -LE memleak")
+    } else {
+        ctest("MemCheck -LE memleak||kdbtests")
+    }
+}
+
+def cnokdbtest() {
+    ctest("Test -LE kdbtests")
+}
+
+def xunitUpload() {
+    step([$class: 'XUnitBuilder',
+          thresholds: [
+            [$class: 'SkippedThreshold', failureThreshold: '0'],
+            [$class: 'FailedThreshold', failureThreshold: '0']
+          ],
+          tools: [
+            [$class: 'CTestType',
+                pattern: 'build/Testing/**/*.xml']
+        ]
+    ])
+}
+
+def image_id(image) {
+    def t = checksum(image.file)
+    return "${image.id}:${t}"
+}
+
+def checksum(file) {
+    return sh(returnStdout: true,
+              script: "cat $file | sha256sum | dd bs=1 count=8 status=none")
```

Comment:
weird statement: useless cat and why using sha256sum if most of the 256 bits are discarded? crc32 seems to have the desired length.

Reply:
I took this statement from the i3wm build pipeline. 

While you are technically correct, I would argue that sha256 bins are distributed on every distribution I know of, while for example that my personal system does not ship crc32 bin's anymore.
Hence this is much less taken because it is perfect and more because it is available everywhere,. The added downside of computation time is complete neglectable.

Reply:
Okay, but what is the point of cutting it at all? Is sounds like a dangerous decision to save some bits and have the risk of easily get collisions.

Reply:
> It needs to be unique enough so that a change to the dockerfile immediately is reflected in the hash itself.

Hashes solve this problem. So crc32 actually is not a good choice.

> The length is long enough to guarantee that while not too long to be cumbersome in logs.

E.g. md5 would do? Hashes are not designed to only use parts of it. Systems that show only parts (like git) have the full shas stored, and only show small parts if the small part is actually unique.

> what exactly would be dangerous here? The only thing you can achive (theoretically) is that the dockerfile is modified and the build system does not initiate a rebuild...

I did not know that, some documentation would help. I was afraid that wrong Dockerfiles are used then (or similar).

I only wondered about the code smell. If it is not important anyway, please keep it as is.

Reply:
Any hash function sufficies by definition as small changes in the input need to generate a completely different hash.

Functionality wise it is a quick check to see if the contents of the Dockerfile changes (which is enough in this case as we don't use any other file to generate the images). If it does it triggers a rebuild of the images as the old one can not be found locally or in the registry
https://github.com/ingwinlu/libelektra/blob/92f1cad3919a1135b8215c5dd072e4a1e5aa75c2/scripts/jenkins/Jenkinsfile#L145

I can pad this section with comments as it might not be clear from the functions used

Reply:
> Any hash function sufficies by definition as small changes in the input need to generate a completely different hash.

It might be the rationale for hash functions but I could not find something like that in the [RFC for SHA](https://tools.ietf.org/html/rfc4634):

> it is computationally infeasible to (1) find a message that
>  corresponds to a given message digest, or (2) find two different
>  messages that produce the same message digest.

If you say (and document) that it does not matter, it is okay. But the code smells. I could be wrong though, maybe he/she actually proved that this reduction is something sane to do. I heard that it was very difficult to prove that AES256 is more secure than AES128. md5 might be weak in deliberate attacks but at least its properties are known.

Btw. I was already involved in the bug triage where a hash sum started with a file name which caused http requests to be redirected wrongly.

Reply:
Btw. why not simply *use* the full hash and *log* the first 8 bytes.

This might lead to a small confusion in the logs but at least it will not bring us in a situation where docker images are not rebuilt and nobody knows why.

Reply:
I will use full length hashes